lib/sse_consumer.ex

```diff
@@ -0,0 +1,150 @@
+            receive do
+              %HTTPoison.AsyncHeaders{id: ^conn_id, headers: _} ->
+                Logger.info("SSEConsumer connected to `#{url}`")
+                {:noreply, state}
+              after
+                100 ->
+                  Logger.warn("SSEConsumer timed out waiting for the headers")
+                  disconnect_and_die(state, :headers_timeout)
+            end
+          %HTTPoison.AsyncStatus{id: ^conn_id, code: 400} ->
+            receive do
+              %HTTPoison.AsyncHeaders{headers: _} ->
+                Logger.warn("SSEConsumer received HTTP 400")
+                disconnect_and_die(state, :http_400)
+            end
+          after
+            2_000 ->
+              Logger.warn("SSEConsumer timed out waiting for a response")
+              disconnect_and_die(state, :timeout)
+        end
+      {:error, %HTTPoison.Error{reason: reason}} ->
+        Logger.warn("SSEConsumer failed to connect. reason: `#{reason}`")
+        disconnect_and_die(state, {:httpoison, reason})
+    end
+  end
+
+  #---------------------------------------------------------------------------
+  # Handling HTTPoison messages
+  #---------------------------------------------------------------------------
+
+  def handle_info(%HTTPoison.AsyncChunk{chunk: ""}, %State{} = state) do
+    {:noreply, state}
+  end
+
+  def handle_info(%HTTPoison.AsyncEnd{}, %State{} = state) do
+    disconnect_and_die(state, :finished)
+  end
+
+  def handle_info(%HTTPoison.AsyncChunk{chunk: chunk}, %State{} = state) do
+    chunk = state.remaining_stream <> chunk
+    case ServerSentEvent.parse_all(chunk) do
+      {:ok, {events, remaining_stream}} ->
+        handle_events(events, state)
+        {:noreply, %{state | remaining_stream: remaining_stream}}
+      {:error, reason} ->
+        Logger.warn("SSE Consumer failed to parse chunk `#{chunk}` because of `#{reason}`")
+        disconnect_and_die(state, {:sse_parse, reason})
+    end
+  end
+
+  def handle_info(%HTTPoison.Error{reason: reason}, %State{} = state) do
+    Logger.warn("SSE Consumer streaming error; #{inspect(reason)}")
+    disconnect_and_die(state, {:httpoison, reason})
+  end
+
+  def handle_info(info, state) do
+    # If connection is established after timeout we need to drop messages streamed to process via HTTPoison
+    Logger.warn("SSE Consumer: unexpected info message: #{inspect info}")
+    {:noreply, state}
+  end
+
+  #===========================================================================
+  # Utility functions
+  #===========================================================================
+
+  defp handle_events(events, state) do
+    message = {:sse, self(), events}
+    send(state.recipient, message)
+  end
+
+
+  defp stream_opts, do: [stream_to: self(), timeout: 50_000, recv_timeout: 50_000]
+
+
+  defp disconnect_and_die(state = %State{conn_id: conn_id}, reason) do
+    send(state.recipient, {:sse_disconnected, self(), reason})
```

Comment:
Should the caller also monitor this process as there is no guarantee that this disconect message will be sent

Reply:
so I think this is ok. As long as the recipient process is linked then they will die together for any unexpected failure.

Reply:
I don't think they need to. If the consumer exits normally, it should report `{:sse_disconnected, ..., ...}`. If it doesn't the caller will know about it, because they must have* started it with `start_link` and they will get the exit signal.

*they technically didn't but if they circumvented the API, it's their call ;)